internal/controller/managedcluster_controller.go

```diff
@@ -683,6 +695,357 @@ func (r *ManagedClusterReconciler) machinesAvailable(ctx context.Context, namesp
+	}
+
+	vsphereClIdty := &capv.VSphereClusterIdentity{}
+	if err := r.Client.Get(ctx, client.ObjectKey{
+		Name: vsphereCluster.Spec.IdentityRef.Name,
+	}, vsphereClIdty); err != nil {
+		return fmt.Errorf("failed to get VSphereClusterIdentity %s: %s", vsphereCluster.Spec.IdentityRef.Name, err)
+	}
+
+	vsphereSecret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, client.ObjectKey{
+		Name:      vsphereClIdty.Spec.SecretName,
+		Namespace: r.SystemNamespace,
+	}, vsphereSecret); err != nil {
+		return fmt.Errorf("failed to get VSphere Secret %s: %s", vsphereClIdty.Spec.SecretName, err)
+	}
+
+	vsphereMachines := &capv.VSphereMachineList{}
+	if err := r.Client.List(
+		ctx,
+		vsphereMachines,
+		&client.ListOptions{
+			Namespace: managedCluster.Namespace,
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				hmc.ClusterNameLabelKey: managedCluster.Name,
+			}),
+			Limit: 1,
+		},
+	); err != nil {
+		return fmt.Errorf("failed to list VSphereMachines for cluster %s: %s", managedCluster.Name, err)
+	}
+	ccmSecret, ccmConfig, err := generateVSphereCCMConfigs(vsphereCluster, vsphereSecret, &vsphereMachines.Items[0])
+	if err != nil {
+		return fmt.Errorf("failed to generate VSphere CCM config: %s", err)
+	}
+	csiSecret, err := generateVSphereCSISecret(vsphereCluster, vsphereSecret, &vsphereMachines.Items[0])
+	if err != nil {
+		return fmt.Errorf("failed to generate VSphere CSI secret: %s", err)
+	}
+
+	if err := applyCCMConfigs(ctx, kubeconfSecret, ccmSecret, ccmConfig, csiSecret); err != nil {
+		return fmt.Errorf("failed to apply VSphere CCM/CSI secrets: %s", err)
+	}
+
+	return nil
+}
+
+func generateVSphereCCMConfigs(vCl *capv.VSphereCluster, vScrt *corev1.Secret, vMa *capv.VSphereMachine) (*corev1.Secret, *corev1.ConfigMap, error) {
+	ns := "kube-system"
+	secretName := "vsphere-cloud-secret"
+	secretData := map[string][]byte{
+		fmt.Sprintf("%s.username", vCl.Spec.Server): vScrt.Data["username"],
+		fmt.Sprintf("%s.password", vCl.Spec.Server): vScrt.Data["password"],
+	}
+	ccmCfg := map[string]any{
+		"global": map[string]any{
+			"port":            443,
+			"insecureFlag":    true,
+			"secretName":      secretName,
+			"secretNamespace": ns,
+		},
+		"vcenter": map[string]any{
+			vCl.Spec.Server: map[string]any{
+				"server": vCl.Spec.Server,
+				"datacenters": []string{
+					vMa.Spec.Datacenter,
+				},
+			},
+		},
+		"labels": map[string]any{
+			"region": "k8s-region",
+			"zone":   "k8s-zone",
+		},
+	}
+	ccmCfgYaml, err := yaml.Marshal(ccmCfg)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to marshal CCM config: %s", err)
+	}
+	cmData := map[string]string{
+		"vsphere.conf": string(ccmCfgYaml),
+	}
+	return makeSecret(secretName, ns, secretData),
+		makeConfigMap("cloud-config", ns, cmData),
+		nil
+}
+
+func generateVSphereCSISecret(vCl *capv.VSphereCluster, vScrt *corev1.Secret, vMa *capv.VSphereMachine) (*corev1.Secret, error) {
+	csiCfg := `
+[Global]
+cluster-id = "{{ .ClusterID }}"
```

Comment:
I'm wondering, don't we need to set an empty `thumbprint` here explicitly? I've seen such a setting here and there. And the same question goes regarding the CPI configuration in the `global` section

Reply:
It's optional see [CSI docs](https://docs.vmware.com/en/VMware-vSphere-Container-Storage-Plug-in/2.0/vmware-vsphere-csp-getting-started/GUID-BFF39F1D-F70A-4360-ABC9-85BDAFBE8864.html).
The CAPV made it mandatory in their configs for some unknown reasons